Added unit tests for HabitFactSpecification

Shcherba04 opened the pull request:
toPredicate Id, HabitIdwithJoin, Content, ContentNull testing, MultipleCriteria

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Tests**
  * Added comprehensive unit tests to improve filtering and search functionality for habits, ensuring accurate results for various search conditions.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Shcherba04 committed:
Added tests for toPredicate Id, HabitIdwithJoin, Content, ContentNull testing, MultipleCriteria

natri0 commented on service/src/test/java/greencity/filters/HabitFactSpecificationTest.java:
```diff
@@ -0,0 +1,236 @@
+class HabitFactSpecificationTest {
+
+    @Mock
+    private Root<HabitFact> HabitFactRoot;
```
fields should be in camelCase like rabbit suggested
```suggestion
    @Mock
    private Root<HabitFact> habitFactRoot;
```

natri0 commented on service/src/test/java/greencity/filters/HabitFactSpecificationTest.java:
```diff
@@ -0,0 +1,236 @@
+    @Mock
+    private CriteriaBuilder builder;
+
+    @Mock
+    private CriteriaQuery MockCriteriaQuery;
```
same as above
```suggestion
    @Mock
    private CriteriaQuery mockCriteriaQuery;
```

natri0 commented on service/src/test/java/greencity/filters/HabitFactSpecificationTest.java:
```diff
@@ -0,0 +1,236 @@
+
+    @Test
+    void toPredicate_Id(){
+        List<SearchCriteria> searchCriteriaList = new ArrayList<>();
+        searchCriteriaList.add(SearchCriteria.builder()
+                .key("id")
+                .type("id")
+                .value(2L)
+                .build());
```
minor nitpick: use `List.of`
```suggestion
        List<SearchCriteria> searchCriteriaList = List.of(
                SearchCriteria.builder()
                        .key("id")
                        .type("id")
                        .value(2L)
                        .build());
```

natri0 commented on service/src/test/java/greencity/filters/HabitFactSpecificationTest.java:
```diff
@@ -0,0 +1,236 @@
+    @Test
+    @MockitoSettings(strictness = Strictness.LENIENT)
+    void toPredicate_HabitIdWithJoin() {
+        List<SearchCriteria> searchCriteriaList = new ArrayList<>();
+        searchCriteriaList.add(SearchCriteria.builder()
+                .key("habitId")
+                .type("habitId")
+                .value(5L)
+                .build());
```
same as above, use List.of

natri0 commented on service/src/test/java/greencity/filters/HabitFactSpecificationTest.java:
```diff
@@ -0,0 +1,236 @@
+                .key("content")
+                .type("content")
+                .value("Smth")
+                .build());
```
same as above, use List.of

natri0 commented on service/src/test/java/greencity/filters/HabitFactSpecificationTest.java:
```diff
@@ -0,0 +1,236 @@
+                .key("content")
+                .type("content")
+                .value(" ")
+                .build());
```
same as above, use List.of

natri0 commented on service/src/test/java/greencity/filters/HabitFactSpecificationTest.java:
```diff
@@ -0,0 +1,236 @@
+
+    @Test
+    void toPredicate_multipleCriteria(){
+        List<SearchCriteria> searchCriteriaList = new ArrayList<>();
+        searchCriteriaList.add(SearchCriteria.builder()
+                .key("id")
+                .type("id")
+                .value(2L)
+                .build());
+        searchCriteriaList.add(SearchCriteria.builder()
+                .key("habitId")
+                .type("habitId")
+                .value(5L)
+                .build());
+        searchCriteriaList.add(SearchCriteria.builder()
+                .key("content")
+                .type("content")
+                .value("Smth")
+                .build());
```
same as above, use List.of

natri0 commented on service/src/test/java/greencity/filters/HabitFactSpecificationTest.java:
```diff
@@ -0,0 +1,236 @@
+    @Mock
+    private CriteriaQuery MockCriteriaQuery;
+
+    @Mock
+    private Path<Object> id;
+
+    @Mock
+    private Path<Long> habitIdPath;
+
+    @Mock
+    private Path<Long> habitFactRootIdPath;
+
+    @Mock
+    private Join<HabitFact,Habit> habitJoin;
+
+    @Mock
+    private Path<String> contentPath;
+
+    @Mock
+    private Predicate likePredicate;
+
+    @Mock
+    private Expression<String> as;
+    HabitFactSpecification specification;
```
for some predicates/paths/etc you use local variables, but for some others you use fields; maybe use local variables for everything, or at least document which fields correspond to which test?

Shcherba04 replied:
Due lack of time, I can only document it.

natri0 reviewed (changes requested):
some minor nitpicks, but is good otherwise.

Kostya-Pav commented on service/src/test/java/greencity/filters/HabitFactSpecificationTest.java:
```diff
@@ -0,0 +1,236 @@
+
+    @Test
+    void toPredicate_multipleCriteria(){
+        List<SearchCriteria> searchCriteriaList = new ArrayList<>();
+        searchCriteriaList.add(SearchCriteria.builder()
+                .key("id")
+                .type("id")
+                .value(2L)
+                .build());
+        searchCriteriaList.add(SearchCriteria.builder()
+                .key("habitId")
+                .type("habitId")
+                .value(5L)
+                .build());
+        searchCriteriaList.add(SearchCriteria.builder()
+                .key("content")
+                .type("content")
+                .value("Smth")
+                .build());
+        specification=new HabitFactSpecification(searchCriteriaList);
```
You already use this logic multiple times. It’s better to move the shared setup into a @BeforeAll method to avoid duplication when preparing test objects.

Shcherba04 replied:
If I use @BeforeAll or @BeforeEach, code need full rewrite. And I preparing to test for master degree, which would be on 08.07. So I can't implement this setup:(

Shcherba04 committed:
Update like in suggestion

Co-authored-by: natri0 <dev@natri.fyi>

Shcherba04 committed:
Another like in camel case

Co-authored-by: natri0 <dev@natri.fyi>

Shcherba04 committed:
List.of without creating new arraylist<>

Co-authored-by: natri0 <dev@natri.fyi>

Shcherba04 committed:
List.of not new arraylist

Shcherba04 committed:
Update HabitFactSpecificationTest.java

Comments

khomenkooleksandr commented on service/src/test/java/greencity/filters/HabitFactSpecificationTest.java:
```diff
@@ -0,0 +1,241 @@
+                .value(5L)
+                .build());
+        specification = new HabitFactSpecification(searchCriteriaList);
+        when(HabitFactRoot.join(HabitFact_.habit)).thenReturn(habitJoin);
```
HaBitFactRoot  code convention habitRactRoot

khomenkooleksandr commented on service/src/test/java/greencity/filters/HabitFactSpecificationTest.java:
```diff
@@ -0,0 +1,241 @@
+
+    //Translation join 
+    @Mock
+    private Predicate likePredicate;
```
variable is not used

khomenkooleksandr commented on service/src/test/java/greencity/filters/HabitFactSpecificationTest.java:
```diff
@@ -0,0 +1,241 @@
+    private Predicate likePredicate;
+
+    @Mock
+    private Expression<String> as;
```
variable is not use

khomenkooleksandr commented on service/src/test/java/greencity/filters/HabitFactSpecificationTest.java:
```diff
@@ -0,0 +1,241 @@
+        Predicate mockPredicate = mock(Predicate.class);
+        when(builder.equal(habitIdPath, 5L)).thenReturn(mockPredicate);
+        Predicate conjunctionPredicate = mock(Predicate.class);
+        when(builder.conjunction()).thenReturn(conjunctionPredicate);
```
Code Duplication:
The test cases repeat similar setup logic for mocking builder.conjunction(), builder.and(), etc. This could be refactored into a common setup method  @BeforeEach block to reduce duplication.

Shcherba04 replied:
To your comment, HabitFactSpecification id and habitid doesn't cover null values. So I added tests and they will have errors due that.

khomenkooleksandr commented on service/src/test/java/greencity/filters/HabitFactSpecificationTest.java:
```diff
@@ -0,0 +1,241 @@
+                .key("content")
+                .type("content")
+                .value("Smth")
+                .build());
```
Hardcoded Values:
The test data ( 2L for id, 5L for habitId, "Smth" for content etc) is hardcoded. Using constants

khomenkooleksandr commented on service/src/test/java/greencity/filters/HabitFactSpecificationTest.java:
```diff
@@ -0,0 +1,241 @@
+        verify(builder).and(secondCombination, contentAndPredicate);
+
+        }
+
+
```
extra spaces

khomenkooleksandr commented on service/src/test/java/greencity/filters/HabitFactSpecificationTest.java:
```diff
@@ -0,0 +1,241 @@
+        when(habitFactTranslationRoot.get(HabitFactTranslation_.habitFact).get(HabitFact_.id)).thenReturn(habitIdPath);
+        when(HabitFactRoot.get(HabitFact_.id)).thenReturn(habitFactRootIdPath);
+
+        Predicate likePredicate = mock(Predicate.class);
```
why don't you use an already declared variable?

khomenkooleksandr commented on service/src/test/java/greencity/filters/HabitFactSpecificationTest.java:
```diff
@@ -0,0 +1,241 @@
+    }
+
+    @Test
+    void toPredicate_ContentNull(){
```
toPredicateContentEmpty

khomenkooleksandr reviewed (changes requested):
The test class is robust and covers the main functionality of HabitFactSpecification. By addressing the suggestions above—simplifying test logic, improving naming, adding edge cases, and reducing redundancy—you can make the tests more maintainable and easier to understand.
Missing Edge Cases:
The tests do not cover invalid or null SearchCriteria values (except for the empty content case). 
You checked only in method toPredicate_ContentNull empty content.
Adding tests for these edge cases would improve robustness.

Shcherba04 committed:
Updated

Deleted unused variable
Deleted declaration of likePredicate in join test and multipleCriteria
added constants
camelCase for habitFactRoot

Shcherba04 committed:
Null SearchCriteria values for id and habitid
